File: season02/quest04/ex05/my_spaceship.c
#include <stdlib.h>
#include <stdio.h>

struct Vector
{
    int x;
    int y;
};

struct Direction
{
    char* name;
    struct Vector vector;
};

struct Direction directions[4] = {
    {"up", {0, -1} },
    {"right", {1, 0} },
    {"down", {0, 1} },
    {"left", {-1, 0} },
};

void vector_add(struct Vector * d, struct Vector s) {
    d->x += s.x;
    d->y += s.y;
}


int direction_direction_id(int direction_id) {
    direction_id = direction_id % 4;
    direction_id += 4;
    direction_id = direction_id % 4;
    return direction_id;
}


char* my_spaceship(char* param_1)
{
    struct Vector position = {0, 0};
    int direction_id = 0;
    char*p = param_1;
    while(*p!='\0') {
        switch(*p) {
            case 'L':
                direction_id -= 1;
                direction_id = direction_direction_id(direction_id);
                break;
            case 'R':
                direction_id += 1;
                direction_id = direction_direction_id(direction_id);
                break;
            case 'A':
                vector_add(&position, directions[direction_id].vector);
                break;
        }
        ++p;
    }
    char* result = (char*) malloc(100);
    sprintf(result, "{x: %d, y: %d, direction: '%s'}", position.x, position.y, directions[direction_id].name);

    return result;
}

// int main() {
//     printf("%s", my_spaceship("RAALALL"));
// }

File: season01/my_spaceship/ex00/my_spaceship.js
class Vector {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }
    add(another) {
        this.x += another.x;
        this.y += another.y;
    }
}

const directions = {
    up: new Vector(0, -1),
    right: new Vector(1, 0),
    down: new Vector(0, 1),
    left: new Vector(-1, 0),
};
const direction_names = Object.keys(directions);
const directions_length = direction_names.length;

class Direction {
    constructor(direction_index) {
        this.direction_index = direction_index;
        this._direction()
    }
    _direction() {
        if(this.direction_index<0) {
            this.direction_index += directions_length;
        }
        this.direction_index = this.direction_index % directions_length;
        this.direction_name = direction_names[this.direction_index];
        this.direction = directions[this.direction_name];
    }
    rotate_right() {
        this.direction_index += 1;
        this._direction()
    }
    rotate_left() {
        this.direction_index -= 1;
        this._direction()
    }
}
function my_spaceship(param_1) {
    let position = new Vector(0, 0);
    let direction = new Direction(0);
    for (const val of param_1) {
        switch (val) {
            case 'L':
                direction.rotate_left();
                break;
            case 'R':
                direction.rotate_right();
                break;
            case 'A':
                position.add(direction.direction);
                break;
            default:
                break;
        }
    }
    return `{x: ${position.x}, y: ${position.y}, direction: '${direction.direction_name}'}`
    //return {x:position.x, y:position.y, direction:direction.direction_name}
}